python/hsfs/core/arrow_flight_client.py

```diff
@@ -203,51 +208,82 @@ def read_path(self, path):
     def is_flyingduck_query_object(self, query_obj):
         return isinstance(query_obj, dict) and "query_string" in query_obj
 
-    def create_query_object(self, query, query_str):
+    def create_query_object(self, query, query_str, on_demand_fg_aliases=[]):
         features = {}
+        connectors = {}
         for fg in query.featuregroups:
             fg_name = self._serialize_featuregroup_name(fg)
+            fg_connector = self._serialize_featuregroup_connector(
+                fg, query, on_demand_fg_aliases
+            )
             features[fg_name] = [feat.name for feat in fg.features]
-        filters = self._serialize_filter_expression(query)
-        for feature in features:
-            features[feature] = list(features[feature])
+            connectors[fg_name] = fg_connector
+        filters = self._serialize_filter_expression(query.filters, query)
+        for fg_name in features:
+            features[fg_name] = list(features[fg_name])
 
         query = {
             "query_string": self._translate_to_duckdb(query, query_str),
             "features": features,
             "filters": filters,
+            "connectors": connectors,
         }
         return query
 
+    def _serialize_featuregroup_connector(self, fg, query, on_demand_fg_aliases):
+        connector = {}
+        if isinstance(fg, feature_group.ExternalFeatureGroup):
+            connector["type"] = fg.storage_connector.type
+            connector["options"] = fg.storage_connector.snowflake_connector_options()
+            connector["query"] = fg.query
+            for on_demand_fg_alias in on_demand_fg_aliases:
+                if on_demand_fg_alias.on_demand_feature_group.name == fg.name:
+                    connector["alias"] = on_demand_fg_alias.alias
+                    break
+            if query._left_feature_group == fg:
+                connector["filters"] = self._serialize_filter_expression(
+                    query._filter, query, True
+                )
+            else:
+                for join_obj in query._joins:
+                    if join_obj._query._left_feature_group == fg:
+                        connector["filters"] = self._serialize_filter_expression(
+                            join_obj._query._filter, join_obj._query, True
+                        )
+        else:
+            connector["type"] = "hudi"
+
+        return connector
+
     def _serialize_featuregroup_name(self, fg):
         return f"{fg._get_project_name()}.{fg.name}_{fg.version}"
 
-    def _serialize_filter_expression(self, query):
-        if query.filters is None:
+    def _serialize_filter_expression(self, filters, query, short_name=False):
```

Comment:
why was it changed? To me it looks like previously it was better since now we always have to extract `query._filter, query` while we could just pass `query`

Reply:
Since we have multiple different sources now (snowflake, hudi), we need to handle pushdowns per featuregroup. If you have a join between a hudi and a snowflake table, we need to push down the filter that applies to the snowflake table (s. https://github.com/logicalclocks/feature-store-api/pull/1069/files#diff-a081d1539ddd7fe9470d2077edfc0c45f48fd9db17d078799ac14c952c7e0df7R221). I still wanted to re-use that method though.

Reply:
Indeed, I think we change it back, I think it was some intermediate state where i needed it like this, but looks like it can be reverted again.